libraries/chain/include/eosio/chain/merkle.hpp

```diff
@@ -1,29 +1,82 @@
+   return digest_type::hash(std::make_pair(std::cref(a), std::cref(b)));
+}
 
-   inline auto make_canonical_pair(const digest_type& l, const digest_type& r) {
-      return make_pair(make_canonical_left(l), make_canonical_right(r));
-   };
+// does not overwrite passed sequence
+//
+// log2 recursion OK, uses less than 5KB stack space for 4 billion digests
+// appended (or 0.25% of default 2MB thread stack size on Ubuntu).
+// -----------------------------------------------------------------------
+template <class It, bool async = false>
+requires std::is_same_v<std::decay_t<typename std::iterator_traits<It>::value_type>, digest_type>
+inline digest_type calculate_merkle_pow2(const It& start, const It& end) {
+   auto size = end - start;
+   assert(size >= 2);
+   assert(detail::bit_floor(static_cast<size_t>(size)) == size);
 
-   /**
-    *  Calculates the merkle root of a set of digests, if ids is odd it will duplicate the last id.
-    *  Uses make_canonical_pair which before hashing sets the first bit of the previous hashes
-    *  to 0 or 1 to indicate the side it is on.
-    */
-   digest_type legacy_merkle( deque<digest_type>&& ids );
+   if (size == 2)
+      return hash_combine(start[0], start[1]);
+   else {
+      if (async && size >= 4096) {                    // below 4096, starting async threads is overkill
+         std::array<std::future<digest_type>, 4> fut; // size dictates the number of threads (must be power of two)
+         size_t slice_size = size / fut.size();
 
-   /**
-    * Calculates the merkle root of a set of digests. Does not manipulate the digests.
-    */
-   digest_type calculate_merkle( deque<digest_type>&& ids );
+         for (size_t i=0; i<fut.size(); ++i)
+            fut[i] = std::async(std::launch::async, calculate_merkle_pow2<It>,
+                                start + slice_size * i, start + slice_size * (i+1));
 
-} } /// eosio::chain
+         std::array<digest_type, fut.size()> res;
+         for (size_t i=0; i<fut.size(); ++i)
+            res[i] = fut[i].get();
+
+         return calculate_merkle_pow2(res.begin(), res.end());
+      } else {
+         auto mid = start + size / 2;
+         return hash_combine(calculate_merkle_pow2(start, mid), calculate_merkle_pow2(mid, end));
+      }
+   }
+}
+
+template <class It>
+requires std::is_same_v<std::decay_t<typename std::iterator_traits<It>::value_type>, digest_type>
+inline digest_type calculate_merkle(const It& start, const It& end) {
+   assert(end >= start);
+   auto size = static_cast<size_t>(end - start);
+   if (size <= 1)
+      return (size == 0) ? digest_type{} : *start;
+
+   auto midpoint = detail::bit_floor(size);
+   if (size == midpoint)
+      return calculate_merkle_pow2<It, true>(start, end);
+
+   auto mid = start + midpoint;
+   return hash_combine(calculate_merkle_pow2<It, true>(start, mid), calculate_merkle(mid, end));
+}
+
+}
+
+inline digest_type calculate_merkle(const deque<digest_type>& ids) {
```

Comment:
Seems this should be a template so it could take any range that provides `cbegin(), cend()`.

Reply:
done.